### ISSUE TITLE
[8.0.x] New IO - Potential deadlock

### ISSUE THREAD

**Issue by fwbrasil** [scalaz/scalaz#1665]
@jdegoes I'm trying to benchmark the new IO implementation but it seems that it has a concurrency issue. This code runs for some time but at some point, it blocks indefinitely:

```scala
import scalaz._
import scalaz.effect.IO
import java.util.concurrent.Executors
import scalaz.effect.RTS

object Bug extends App {
  val e = Executors.newSingleThreadExecutor()
  object rts extends RTS
  import rts._

  for (i <- (0 until 10000)) {
    println(i)
    val t = IO.async[Int] { cb =>
      try e.submit[Unit](() => cb(\/-(1)))
      catch {
        case t: Throwable => t.printStackTrace()
      }
    }
    unsafePerformIO(t)
  }

  e.shutdown()
}
```

Thread dump:
```
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.144-b01 mixed mode):

"Attach Listener" #11 daemon prio=9 os_prio=0 tid=0x00007f464c001000 nid=0x4e3b waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"pool-1-thread-1" #10 prio=5 os_prio=0 tid=0x00007f46a893b000 nid=0x4e0d waiting on condition [0x00007f46766a8000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000071c561520> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"Service Thread" #9 daemon prio=9 os_prio=0 tid=0x00007f46a80e6800 nid=0x4e0a runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread3" #8 daemon prio=9 os_prio=0 tid=0x00007f46a80db800 nid=0x4e09 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread2" #7 daemon prio=9 os_prio=0 tid=0x00007f46a80d9000 nid=0x4e08 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #6 daemon prio=9 os_prio=0 tid=0x00007f46a80d7800 nid=0x4e07 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=0 tid=0x00007f46a80d4800 nid=0x4e06 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 tid=0x00007f46a80d3000 nid=0x4e05 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x00007f46a80a0000 nid=0x4e04 in Object.wait() [0x00007f46771f0000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000071b608ec8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x000000071b608ec8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x00007f46a809b000 nid=0x4e03 in Object.wait() [0x00007f46772f1000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000071b606b68> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x000000071b606b68> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"main" #1 prio=5 os_prio=0 tid=0x00007f46a800b800 nid=0x4df9 waiting on condition [0x00007f46af082000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000071ca73ea0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at scalaz.effect.RTS.tryUnsafePerformIO(RTS.scala:57)
	at scalaz.effect.RTS.tryUnsafePerformIO$(RTS.scala:32)
	at bug.Bug$rts$.tryUnsafePerformIO(Bug.scala:10)
	at scalaz.effect.RTS.unsafePerformIO(RTS.scala:27)
	at scalaz.effect.RTS.unsafePerformIO$(RTS.scala:27)
	at bug.Bug$rts$.unsafePerformIO(Bug.scala:10)
	at bug.Bug$.$anonfun$new$1(Bug.scala:21)
	at bug.Bug$$$Lambda$10/455896770.apply$mcII$sp(Unknown Source)
	at scala.runtime.java8.JFunction1$mcII$sp.apply(JFunction1$mcII$sp.java:12)
	at scala.collection.immutable.Range.foreach(Range.scala:156)
	at bug.Bug$.delayedEndpoint$bug$Bug$1(Bug.scala:13)
	at bug.Bug$delayedInit$body.apply(Bug.scala:8)
	at scala.Function0.apply$mcV$sp(Function0.scala:34)
	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App.$anonfun$main$1$adapted(App.scala:76)
	at scala.App$$Lambda$5/1072408673.apply(Unknown Source)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.App.main(App.scala:76)
	at scala.App.main$(App.scala:74)
	at bug.Bug$.main(Bug.scala:8)
	at bug.Bug.main(Bug.scala)

"VM Thread" os_prio=0 tid=0x00007f46a8093800 nid=0x4e02 runnable 

"GC task thread#0 (ParallelGC)" os_prio=0 tid=0x00007f46a8021000 nid=0x4dfa runnable 

"GC task thread#1 (ParallelGC)" os_prio=0 tid=0x00007f46a8023000 nid=0x4dfb runnable 

"GC task thread#2 (ParallelGC)" os_prio=0 tid=0x00007f46a8024800 nid=0x4dfc runnable 

"GC task thread#3 (ParallelGC)" os_prio=0 tid=0x00007f46a8026800 nid=0x4dfd runnable 

"GC task thread#4 (ParallelGC)" os_prio=0 tid=0x00007f46a8028000 nid=0x4dfe runnable 

"GC task thread#5 (ParallelGC)" os_prio=0 tid=0x00007f46a802a000 nid=0x4dff runnable 

"GC task thread#6 (ParallelGC)" os_prio=0 tid=0x00007f46a802b800 nid=0x4e00 runnable 

"GC task thread#7 (ParallelGC)" os_prio=0 tid=0x00007f46a802d800 nid=0x4e01 runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x00007f46a80e9800 nid=0x4e0b waiting on condition 

JNI global references: 417

```

**Comment by jdegoes** [comment]
@fwbrasil Thanks for the bug report! I'll take a look at this within the next week.

**Comment by fwbrasil** [comment]
Hi @jdegoes, would you have an update on this? I'm planning to publish some benchmark results this weekend and it'd be nice to have the scalaz numbers

**Comment by jdegoes** [comment]
@fwbrasil Sorry about the delay. I pushed this back until after I completed a very large refactoring.

I'm done with the refactoring and taking a look at this next.

Note that `unsafePerformIO` in Scalaz 8 IO is not intended to be called until the "end of the world". So while this example should work and not deadlock, the library is not designed for this style of development (rather, you'd do the for "loop" inside `IO` and at the end of the test, perform `unsafePerformIO` a single time).

**Comment by jdegoes** [comment]
This is now a regression test:

```scala
  def testDeadlockRegression = {
    import scalaz._
    import java.util.concurrent.Executors
    import scalaz.effect.RTS

    val e = Executors.newSingleThreadExecutor()

    for (i <- (0 until 10000)) {
      val t = IO.async[Void, Int] { cb =>
        val _ = e.submit[Unit](() => cb(ExitResult.Completed(1)))
      }
      unsafePerformIO(t)
    }

    e.shutdown() must_=== (())
  }
```

**Comment by fwbrasil** [comment]
Thank you @jdegoes!

**Comment by jdegoes** [comment]
@fwbrasil Sure, thanks for the report!
